src/runtime/impure/lfortran_intrinsics.c

```diff
@@ -649,12 +649,20 @@ LFORTRAN_API char* _lfortran_str_copy(char* s, int32_t idx1, int32_t idx2) {
     if(idx2 <= 0) {
         idx2 = s_len + idx2;
     }
-    char* dest_char = (char*)malloc(idx2-idx1+2);
-    for (int i=idx1; i <= idx2; i++)
-    {
-        dest_char[i-idx1] = s[i-1];
+    char* dest_char;
+    if (idx3 > 0 && (idx1 <= idx2)) {
+        dest_char = (char*)malloc(idx2-idx1+2);
+        for (int i=idx1; i <= idx2; i+=idx3){
+            dest_char[((i - idx1)/(idx3))] = s[i-1];
+        }
+        dest_char[idx2-idx1+1] = '\0';
+    } else if (idx3 < 0 && (idx1 > idx2)) {
```

Comment:
I think we might not need to implement `idx3 < 0` separately. Basically its just an arithmetic progression with start = `idx1`, end < `idx2` and `idx3` as the common difference. So, `idx1 + (number_of_terms - 1) * idx3 < idx2`. The sign doesn't matter here (so it shouldn't in the implementation as well I think).

Anyways upto you.

Reply:
For checking if the arithmetic progression is finite you can simply check if the number of terms coming from the above in-equality is positive. Simple do, `number_of_terms = floor((idx2 - idx1)/idx3 + 1)`. If `number_of_terms` is positive then the result will be finite sized. Otherwise it won't be finite sized. Consider, `idx2 > idx1` and `idx3 < 0` then, `idx2 - idx1 > 0` and hence `number_of_terms <= 0`, so no need to run the loop.

Reply:
If you are doing this for caching efficiency then also it won't matter. Because you need to fill the destination string in the right order. Given that `idx3 < 0` and `idx1 > idx2` If you traverse input string from `idx1` to `idx2` then you will move in the backwards direction (lower cache efficiency for input string) but in forward direction in the output string (higher cache efficiency for output string). If you travel from `idx2` to `idx1` in the input string then you will move in forward direction (higher cache efficiency for input string) but you will be required to fill output string in backward direction (lower cache efficiency). Theoretically, its better to achieve higher cache efficiency in output string (because its smaller and we are taking fixed step size of 1 while filling it i.e., greater spatial locality). In input string it all depends on step size if it is higher then it doesn't matter whether you move in forward or backward directions (indices are already far away so lower spatial locality already).

Reply:
Thanks @czgdp1807 for the insight and suggestions. 
I was aiming to get a structure for working step function in `[start:end:step]` format . A couple of edge cases need to be handled first then I'll address all the suggestions .